### PR TITLE
45 code efficiency improvements

### PR DESCRIPTION
**Pull request by lollcat**
- Introduce `Point` where we save the target and base log probs (and grads) of each point in AIS to prevent unecessary extra forward passes of the target and flow. We can also then use these in the loss estimation.  
- remove `no_grad` from AIS forward pass. This should be implemented by the transition operator if it does not use grad. 
- Don't instantiate `self.annealed_importance_sampler` for losses that don't use AIS in `core.FABModel`
- Rename losses in `core.py` to make them more clear. Add `fab_` to losses that use the AIS + flow bootstrap. `flow_` with samples from the the flow. `target_` with samples from the target.
- Add warning if non-standard loss functions from `core.py` are used. 
- Checked that the GMM and Many Well colab's still work (they are roughly 40% faster after the improvements). 
- For HMC only offer tuning by `p_accept` as this is the only one we currently use. Makes the code simpler.

### CODE REVIEW

File: experiments/aldp/train.py [hunk 1]
```diff
@@ -138,14 +138,14 @@
 
 # Set parameters for training
 ndim = 60
-loss_type = 'alpha_2_div' if 'loss_type' not in config['fab'] \
```
Comment (lollcat) [inline experiments/aldp/train.py:141]: @VincentStimper I'm not super confident about the changes I made to the Aldp scripts - could you please double check these?

File: experiments/aldp/train.py [hunk 2]
```diff
@@ -138,14 +138,14 @@
...
+loss_type = 'fab_p2_over_q_alpha_2_div' if 'loss_type' not in config['fab'] \
         else config['fab']['loss_type']
 transition_type = config['fab']['transition_type']
 flow_type = config['flow']['type']
 
 # Load train data if needed
 lam_fkld = None if not 'lam_fkld' in config['fab'] else config['fab']['lam_fkld']
-if loss_type == 'flow_forward_kl' or lam_fkld is not None:
+if loss_type == 'target_forward_kl' or lam_fkld is not None:
```
Comment (lollcat) [inline experiments/aldp/train.py:148]: Following convention in `core.py`

File: experiments/aldp/train.py [hunk 3]
```diff
@@ -199,9 +199,9 @@
     rb_config = config['training']['replay_buffer']
     if rb_config['type'] == 'uniform':
         def initial_sampler():
-            x, log_w = model.annealed_importance_sampler.sample_and_log_weights(
+            point, log_w = model.annealed_importance_sampler.sample_and_log_weights(
                 batch_size, logging=False)
-            return x, log_w
+            return point.x, log_w
```
Comment (lollcat) [inline experiments/aldp/train.py:204]: AIS now returns an instance `Point` see `fab/sampling_methods/base.py` where this is now defined

File: fab/core.py [hunk 4]
```diff
@@ -104,107 +126,88 @@ def flow_alpha_2_div_nis(self, batch_size: int) -> torch.Tensor:
...
-    def p2_over_q_alpha_2_div(self, batch_size: int) -> torch.Tensor:
+    def fab_p2_over_q_alpha_2_div_inner(self, point: Point, log_w_ais: torch.Tensor) -> \
+            torch.Tensor:
+        return - torch.mean(torch.softmax(log_w_ais, dim=-1) * point.log_q)
+
+    def fab_p2_over_q_alpha_2_div(self, batch_size: int) -> torch.Tensor:
         """Compute the FAB loss with p^2/q as the AIS target."""
         # set ais target distribution to p^2/q
-        self.set_ais_target("p^2/q")
-        if isinstance(self.annealed_importance_sampler.transition_operator, HamiltonianMonteCarlo):
```
Comment (lollcat) [inline fab/core.py:111]: The `create_point` function used within the transition operator, make sure that we only have grads through `log_q(x)` and not through the `x` of the MCMC chain, which means that we do not need to block gradient here anymore.

File: fab/sampling_methods/base.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
+from typing import Tuple, Optional, Union
+import torch
+
+from fab.types_ import LogProbFunc
+
+
+class Point:
```
Comment (lollcat) [inline fab/sampling_methods/base.py:7]: This is the most central change of this PR, and most other changes relate to this.

File: fab/sampling_methods/base.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+def grad_and_value(x, forward_fn):
+    """Calculate the forward pass of a function y = f(x) as well as its gradient w.r.t x."""
+    x = x.detach()
+    x.requires_grad = True
+    y = forward_fn(x)
+    grad = torch.autograd.grad(y, x,  grad_outputs=torch.ones_like(y), retain_graph=True)[0]
+    return grad.detach(), y
+
+
+def create_point(x: torch.Tensor, log_q_fn: LogProbFunc, log_p_fn: LogProbFunc,
```
Comment (lollcat) [inline fab/sampling_methods/base.py:59]: Note: this function blocks grads to everything except log_q(x), as we re-use that in the loss.

File: fab/sampling_methods/base.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def get_intermediate_log_prob(x: Point,
+                              beta: float,
+                              p_sq_over_q_target: bool) -> torch.Tensor:
+    """Get log prob of point according to intermediate AIS distribution.
+    Set AIS final target g=p^2/q if p_sq_over_q_target else g=p.
+    log_prob = (1 - beta) log_q + beta log_g
+    """
+    with torch.no_grad():
```
Comment (lollcat) [inline fab/sampling_methods/base.py:83]: Block grads here as we don't backprop through AIS.

**Review by VincentStimper — APPROVED** [review]
Looks nice and seems to work for the dipeptide as well. I'll keep you posted regarding the performance.

**Comment by VincentStimper** [comment]
Just found a bug, but now it should work.
